tests/array.rs

```diff
@@ -1558,6 +1558,53 @@ fn insert_axis_view() {
     );
 }
 
+#[test]
+fn test_broadcast_shape() {
+    fn test_co<D1, D2>(
+        d1: &D1,
+        d2: &D2,
+        r: Result<<D1 as BroadcastShape<D2>>::BroadcastOutput, ShapeError>,
+    ) where
+        D1: Dimension + BroadcastShape<D2>,
+        D2: Dimension,
+    {
+        let d = d1.broadcast_shape(d2);
+        assert_eq!(d, r);
+    }
+    test_co(&Dim([2, 3]), &Dim([4, 1, 3]), Ok(Dim([4, 2, 3])));
+    test_co(
+        &Dim([1, 2, 2]),
+        &Dim([1, 3, 4]),
+        Err(ShapeError::from_kind(ErrorKind::IncompatibleShape)),
+    );
+    test_co(&Dim([3, 4, 5]), &Ix0(), Ok(Dim([3, 4, 5])));
+    let v = vec![1, 2, 3, 4, 5, 6, 7];
+    test_co(
+        &Dim(vec![1, 1, 3, 1, 5, 1, 7]),
+        &Dim([2, 1, 4, 1, 6, 1]),
+        Ok(Dim(IxDynImpl::from(v.as_slice()))),
+    );
+    let d = Dim([1, 2, 1, 3]);
+    test_co(&d, &d, Ok(d));
+    test_co(
+        &Dim([2, 1, 2]).into_dyn(),
+        &Dim(0),
+        Ok(Dim([2, 1, 2]).into_dyn()),
```

Comment:
Should probably be a shape error when one the rhs is this zero dimension?

```python
>>> import numpy as np
>>> np.arange(4.)
array([0., 1., 2., 3.])
>>> a = np.arange(4.).reshape((2, 1, 2))
>>> a
array([[[0., 1.]],

       [[2., 3.]]])
>>> b = a[:0, :, :0]
>>> b = b.reshape((0, ))
>>> b
array([], dtype=float64)
>>> b.shape
(0,)
>>> a + b
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: operands could not be broadcast together with shapes (2,1,2) (0,)
>>>
```

Reply:
Yes. I did not test the zero dimension in numpy. I will test them and be consistent with numpy.

Reply:
As for zero dimension in numpy, if we do like this:

    import numpy as np
    a = np.arange(4.).reshape((2, 2, 1))
    print(a.shape)
    
    b = a[:2,:2,:0]
    b = b.reshape((0))
    print(b.shape)
    
    print(a+b)

It's just fine:

    (2, 2, 1)
    (0,)
    []

In 1 and 0, numpy will also choose 0：

    import numpy as np
    a = np.arange(2.).reshape((1, 2, 1))
    print(a.shape)
    
    b = a[:0,:2,:1]
    b = b.reshape((0,2,1))
    print(b.shape)
    c=a+b
    print(c.shape)

Output:

    (1, 2, 1)
    (0, 2, 1)
    (0, 2, 1)

Then we can look PyArray_BroadcastToShape method in numpy/numpy/core/src/multiarray/iterators:204:

        for (i = 0; i < PyArray_NDIM(ao); i++, j++) {
            if (PyArray_DIMS(ao)[i] == 1) {
                continue;
            }
            if (PyArray_DIMS(ao)[i] != dims[j]) {
                compat = 0;
                break;
            }
        }
        if (!compat) {
            goto err;
        }

numpy treats 0 values just like other values greater than 1.